src/services/v2/on-chain/mover/debit-card/DebitCardOnChainService.ts

```diff
@@ -0,0 +1,304 @@
+        });
+
+      if (gasLimitObj) {
+        return {
+          error: false,
+          approveGasLimit: '0',
+          unwrapGasLimit: '0',
+          actionGasLimit: this.addGasBuffer(gasLimitObj.toString())
+        };
+      }
+    } catch (error) {
+      addSentryBreadcrumb({
+        type: 'error',
+        category: `${this.sentryCategoryPrefix}.estimateTopUpCompound`,
+        message: 'Failed to estimate top up',
+        data: {
+          error,
+          inputAsset,
+          inputAmount,
+          transferData
+        }
+      });
+
+      throw new OnChainServiceError('Failed to estimate top up').wrap(error);
+    }
+
+    addSentryBreadcrumb({
+      type: 'error',
+      category: `${this.sentryCategoryPrefix}.estimateTopUpCompound`,
+      message: 'Failed to estimate top up: empty gas limit',
+      data: {
+        inputAsset,
+        inputAmount,
+        transferData
+      }
+    });
+
+    throw new OnChainServiceError('Failed to estimate top up: empty gas limit');
+  };
+
+  public topUpCompound = async (
+    inputAsset: SmallTokenInfo,
+    inputAmount: string,
+    transferData: TransferData | undefined,
+    changeStep: (step: LoaderStep) => Promise<void>,
+    actionGasLimit: string,
+    approveGasLimit: string,
+    unwrapGasLimit: string
+  ): Promise<TransactionReceipt> => {
+    const specialTokenHandler = this.specialTokenHandlers.find((h) =>
+      h.canHandle(inputAsset.address, this.network)
+    );
+    if (specialTokenHandler !== undefined) {
+      inputAmount = await specialTokenHandler.unwrap(
+        inputAsset,
+        inputAmount,
+        this.web3Client,
+        this.currentAddress,
+        () => changeStep('Process'),
+        unwrapGasLimit
+      );
+      inputAsset = specialTokenHandler.getUnwrappedToken();
+    }
+
+    if (!sameAddress(inputAsset.address, this.usdcAssetData.address)) {
+      transferData = await this.swapService.getTransferData(
+        this.usdcAssetData.address,
+        inputAsset.address,
+        toWei(inputAmount, inputAsset.decimals),
+        true,
+        getSlippage(inputAsset.address, this.network)
+      );
+
+      const estimation = await this.estimateTopUpCompound(
+        inputAsset,
+        inputAmount,
+        transferData
+      );
+
+      actionGasLimit = estimation.actionGasLimit;
+      approveGasLimit = estimation.approveGasLimit;
+    }
+
+    await changeStep('Confirm');
+
+    return await this.executeTransactionWithApprove(
+      inputAsset,
+      this.centralTransferProxyAddress,
+      inputAmount,
+      async (newGasLimit) =>
+        this.topUp(
+          inputAsset,
+          inputAmount,
+          transferData,
+          () => changeStep('Process'),
+          newGasLimit
+        ),
+      () => this.estimateTopUpCompound(inputAsset, inputAmount, transferData),
+      () => changeStep('Process'),
+      approveGasLimit,
+      actionGasLimit
+    );
+  };
+
+  topUp = async (
```

Comment:
```suggestion
  protected async topUp(
```
Should use `protected` access modifier to enable further extensions if needed by allowing inheritance
Otherwise explicitly set `private`